lib/graphlient/adapters/faraday_adapter.rb

```diff
@@ -0,0 +1,40 @@
+require 'faraday'
+require 'faraday_middleware'
+
+module Graphlient
+  module Adapters
+    class FaradayAdapter
+      attr_accessor :url, :headers
+
+      def initialize(url, headers: {})
+        @url = url
+        @headers = headers
+      end
+
+      def execute(document:, operation_name:, variables:, context:)
+        response = conn.post do |req|
+          context[:headers].each { |k, v| req.headers[k] = v } if context[:headers]
+          req.body = {
+            query: document.to_query_string,
+            operationName: operation_name,
+            variables: variables.to_json
+          }.to_json
+        end
+        response.body
+      rescue Faraday::ClientError => e
+        { 'errors' => [{ 'message' => "#{e.response[:status]} #{e.response[:body]}" }] }
```

Comment:
I've always preferred exceptions to returning a JSON with errors. Do we have strong opinions to doing this vs. checking for `errors` and raising? There're many other exceptions possible other than `Faraday::ClientError` and not raising requires the caller to have both handlers.

If you choose to handle `errors` and raise, then that should be a Faraday middleware and `rescue` can be removed here.

Reply:
it makes sense to switch to Faradary middleware, will do that in separate PR since want to get this PR out the door sooner to unblock this release.

https://github.com/ashkan18/graphlient/issues/16